File: src/lib/site-import/status.ts
/* eslint-disable complexity */
import { ApolloClient, NormalizedCacheObject } from '@apollo/client';
import chalk from 'chalk';
import debugLib from 'debug';
import gql from 'graphql-tag';

import { AppForImport, currentUserCanImportForApp } from './db-file-import';
import { AppQuery, AppQueryVariables } from './status.generated';
import {
	App,
	AppEnvironment,
	AppEnvironmentImportStatus,
	AppEnvironmentStatusProgressStep,
	Job,
	Maybe,
	PrimaryDomainSwitchJob,
} from '../../graphqlTypes';
import API from '../../lib/api';
import * as exit from '../../lib/cli/exit';
import { capitalize, formatEnvironment, getGlyphForStatus } from '../../lib/cli/format';
import { ProgressTracker, StepFromServer, StepStatus } from '../../lib/cli/progress';

const debug = debugLib( 'vip:lib/site-import/status' );

const IMPORT_SQL_PROGRESS_POLL_INTERVAL = 5000;

const IMPORT_SQL_PROGRESS_QUERY = gql`
	query App($appId: Int, $envId: Int) {
		app(id: $appId) {
			environments(id: $envId) {
				id
				isK8sResident
				launched
				jobs(types: "sql_import") {
					id
					type
					completedAt
					createdAt
					progress {
						status
						steps {
							id
							name
							status
						}
					}
				}
				importStatus {
					dbOperationInProgress
					importInProgress
					progress {
						started_at
						steps {
							name
							started_at
							finished_at
							result
							output
						}
						finished_at
					}
				}
			}
		}
	}
`;

export interface ImportSqlCheckStatusInput {
	app: App | AppForImport;
	env: AppEnvironment;
	progressTracker: ProgressTracker;
}

interface GetStatusResponse {
	importStatus?: AppEnvironmentImportStatus | null;
	importJob?: Job | PrimaryDomainSwitchJob | null;
	launched?: boolean | null;
}

async function getStatus(
	api: ApolloClient< NormalizedCacheObject >,
	appId: number,
	envId: number
): Promise< GetStatusResponse > {
	const response = await api.query< AppQuery, AppQueryVariables >( {
		query: IMPORT_SQL_PROGRESS_QUERY,
		variables: { appId, envId },
		fetchPolicy: 'network-only',
	} );
	const environments = response.data.app?.environments ?? [];
	if ( ! environments.length ) {
		throw new Error( 'Unable to determine import status from environment' );
	}
	const environment = environments[ 0 ] ?? {};
	const { importStatus, jobs, launched } = environment;

	if ( ! environment.isK8sResident && ! jobs?.length ) {
		return {};
	}

	const [ importJob ] = jobs ?? [];

	return {
		importStatus,
		importJob,
		launched,
	};
}

interface ImportFailedError {
	inImportProgress: boolean;
	commandOutput: string[] | null;
	error: string;
	// eslint-disable-next-line @typescript-eslint/no-redundant-type-constituents
	stepName: 'import_preflights' | 'importing_db' | 'validating_db' | string; // NOSONAR
	launched: boolean;
}

function getErrorMessage( importFailed: ImportFailedError, launched = false ): string {
	debug( { importFailed } );

	const rollbackMessage = launched
		? ''
		: `Your site is ${ chalk.blue(
				'automatically being rolled back'
		  ) } to the last backup prior to your import job.
`;

	let message = importFailed.error;

	if ( importFailed.inImportProgress ) {
		switch ( importFailed.stepName ) {
			case 'import_preflights':
				message += `
This error occurred prior to the mysql batch script processing of your SQL file.

Your site content was not altered.

If this error persists, please contact support.
`;
				break;

			case 'importing_db':
				message += `
This error occurred during the mysql batch script processing of your SQL file.

${ rollbackMessage }`;
				if ( importFailed.commandOutput ) {
					const commandOutput = ( [] as string[] ).concat( importFailed.commandOutput ).join( ';' );
					message += `
Please inspect your input file and make the appropriate corrections before trying again.
The server said:
> ${ chalk.red( commandOutput ) }
`;
				} else {
					message += 'Please contact support and include this message along with your sql file.';
				}
				break;

			case 'validating_db':
				message += `\nThis error occurred during the post-import validation of the imported data.

${ rollbackMessage }
`;
				if ( importFailed.commandOutput ) {
					const commandOutput = ( [] as string[] ).concat( importFailed.commandOutput ).join( ';' );
					message += `
Please inspect your input file and make the appropriate corrections before trying again.
The server said:
> ${ chalk.red( commandOutput ) }
`;
				} else {
					message += 'Please contact support and include this message along with your sql file.';
				}

				break;
			default:
		}
	}
	return message;
}

export async function importSqlCheckStatus( {
	app,
	env,
	progressTracker,
}: ImportSqlCheckStatusInput ) {
	// Stop printing so we can pass our callback
	progressTracker.stopPrinting();

	// NO `console.log` in this function (until results are final)! It will break the progress printing.
	const api = await API();

	if ( ! currentUserCanImportForApp( app ) ) {
		throw new Error(
			'The currently authenticated account does not have permission to view SQL import status.'
		);
	}
	let createdAt: Maybe< string > | undefined;
	let completedAt: Maybe< string > | undefined;
	let overallStatus: string | StepStatus = 'Checking...';

	const setProgressTrackerSuffix = () => {
		const sprite = getGlyphForStatus( overallStatus, progressTracker.runningSprite );
		const formattedCreatedAt = createdAt
			? `${ new Date( createdAt ).toLocaleString() } (${ createdAt })`
			: 'TBD';
		const formattedCompletedAt =
			createdAt && completedAt
				? `${ new Date( completedAt ).toLocaleString() } (${ completedAt })`
				: 'TBD';
		const exitPrompt = '(Press ^C to hide progress. The import will continue in the background.)';

		let statusMessage;
		switch ( overallStatus ) {
			case 'success':
				statusMessage = `Success ${ sprite } imported data should be visible on your site ${
					env.primaryDomain?.name ?? 'N/A'
				}.`;
				break;
			case 'running':
				if ( progressTracker.allStepsSucceeded() ) {
					statusMessage = `Finishing up... ${ sprite } `;
					break;
				}
			// Intentionally no break to get default case:
			// eslint-disable-next-line no-fallthrough
			default:
				statusMessage = `${ capitalize( overallStatus ) } ${ sprite }`;
		}

		const maybeExitPrompt = `${ overallStatus === 'running' ? exitPrompt : '' }`;
		const jobCreateCompleteTimestamps = `
SQL Import Started: ${ formattedCreatedAt }
SQL Import Completed: ${ formattedCompletedAt }`;
		const maybeTimestamps = [ 'running', 'success', 'failed' ].includes( overallStatus )
			? jobCreateCompleteTimestamps
			: '';
		const suffix = `
=============================================================
Status: ${ statusMessage }
Site: ${ app.name ?? 'N/A' } (${ formatEnvironment( env.type ?? 'N/A' ) })${ maybeTimestamps }
=============================================================
${ maybeExitPrompt }
`;
		progressTracker.suffix = suffix;
	};

	const setSuffixAndPrint = () => {
		setProgressTrackerSuffix();
		progressTracker.print();
	};

	progressTracker.startPrinting( setSuffixAndPrint );

	const getResults = (): Promise< string | Job | PrimaryDomainSwitchJob > =>
		new Promise( ( resolve, reject ) => {
			const checkStatus = async (): Promise< unknown > => {
				let status;
				try {
					status = await getStatus( api, app.id ?? -1, env.id ?? -1 );
				} catch ( error ) {
					return reject( { error } );
				}
				const { importStatus, launched } = status;
				let { importJob } = status;

				let jobStatus;
				let jobSteps = [];
				if ( env.isK8sResident ) {
					// in the future the API may provide this in k8s jobs so account for that.
					// Until then we need to create the importJob from the status object.
					if ( ! importJob ) {
						importJob = {};
						const statusSteps = importStatus?.progress?.steps as
							| AppEnvironmentStatusProgressStep[]
							| undefined;

						// if the progress meta isn't filled out yet, wait until it is.
						if ( ! statusSteps ) {
							return setTimeout( () => {
								void checkStatus();
							}, IMPORT_SQL_PROGRESS_POLL_INTERVAL );
						}

						jobSteps = statusSteps.map( step => {
							return {
								id: step.name,
								name: capitalize( step.name?.replace( /_/g, ' ' ) ),
								status: step.result,
							};
						} );

						if (
							statusSteps.some( ( { result } ) => result === 'failed' ) &&
							! statusSteps.find( ( { name, result } ) => name === 'restore_db' && ! result )
						) {
							jobStatus = 'error';
						} else if ( statusSteps.every( ( { result } ) => result === 'success' ) ) {
							jobStatus = 'success';
							const timestamps = statusSteps
								.map( step => step.finished_at )
								.filter( Boolean ) as number[];
							importJob.completedAt = new Date( Math.max( ...timestamps, 0 ) * 1000 ).toUTCString();
						}

						if ( importStatus?.progress?.started_at ) {
							importJob.createdAt = new Date(
								importStatus.progress.started_at * 1000
							).toUTCString();
						}

						importJob.progress = { status: jobStatus, steps: jobSteps };
					}
				} else if ( ! importJob ) {
					return resolve( 'No import job found' );
				}

				jobStatus = importJob.progress?.status ?? 'unknown';
				jobSteps = importJob.progress?.steps ?? [];
				createdAt = importJob.createdAt;
				completedAt = importJob.completedAt;

				const {
					dbOperationInProgress,
					importInProgress,
					progress: importStepProgress,
				} = importStatus ?? {};

				debug( {
					jobStatus,
					completedAt,
					createdAt,
					dbOperationInProgress,
					importInProgress,
					importStepProgress,
				} );

				let jobCreationTime;
				try {
					jobCreationTime = new Date( createdAt ?? '' ).getTime();
				} catch ( err ) {
					debug( 'Unable to parse createdAt to a Date' );
				}

				let failedImportStep;

				if (
					jobCreationTime &&
					( importStepProgress?.started_at ?? 0 ) * 1000 >= jobCreationTime
				) {
					// The contents of the `import_progress` meta are pertinent to the most recent import job
					failedImportStep = importStepProgress?.steps?.find(
						step =>
							step?.result === 'failed' &&
							1000 * ( step.started_at ?? 0 ) > new Date( createdAt ?? '' ).getTime()
					);
				}

				if ( ! jobSteps.length ) {
					return reject( { error: 'Could not enumerate the import job steps', launched } );
				}

				if ( failedImportStep ) {
					// The server marks the step as a success as per the host action, demote it to 'failed'
					const _jobSteps = [ ...jobSteps ];
					const failedJobStepIndex = _jobSteps.findIndex( step => step?.id === 'import' );
					_jobSteps[ failedJobStepIndex ] = {
						..._jobSteps[ failedJobStepIndex ],
						status: 'failed',
					};
					progressTracker.setStepsFromServer( _jobSteps as unknown as StepFromServer[] );
					overallStatus = 'failed';
					setSuffixAndPrint();

					return reject( {
						inImportProgress: true,
						commandOutput: failedImportStep.output,
						error: 'Import step failed',
						stepName: failedImportStep.name,
						launched,
					} as ImportFailedError );
				}

				progressTracker.setStepsFromServer( jobSteps as unknown as StepFromServer[] );

				setSuffixAndPrint();

				if ( jobStatus === 'error' ) {
					return reject( { error: 'Import job failed', steps: jobSteps, launched } );
				}

				if ( jobStatus !== 'running' && completedAt ) {
					return resolve( importJob );
				}

				overallStatus = 'running';

				setTimeout( () => {
					void checkStatus();
				}, IMPORT_SQL_PROGRESS_POLL_INTERVAL );
			};

			// Kick off the check
			void checkStatus();
		} );

	try {
		const results = await getResults();
		if ( typeof results === 'string' ) {
			overallStatus = results;
		} else {
			overallStatus = results.progress?.status ?? 'unknown';
			// This shouldn't be 'unknown'...what should we do here?
		}

		progressTracker.stopPrinting();

		setProgressTrackerSuffix();

		// Print one final time
		progressTracker.print( { clearAfter: true } );

		// This type of result is not an importing error. e.g. no import job was found
		process.exit( 0 );
	} catch ( importFailed ) {
		progressTracker.stopPrinting();
		progressTracker.print( { clearAfter: true } );
		exit.withError(
			getErrorMessage(
				importFailed as ImportFailedError,
				( importFailed as ImportFailedError ).launched
			)
		);
	}
}

export default {
	importSqlCheckStatus,
};

File: internal/siteimport/status.go
package siteimport

import (
	"context"
	"errors"
	"strings"
	"time"

	"github.com/fatih/color"

	"github.com/Automattic/vip/internal/tui"
)

// DefaultPollInterval — IMPORT_SQL_PROGRESS_POLL_INTERVAL (status.ts:25).
const DefaultPollInterval = 5 * time.Second

// JobStep is one step of the import job as reported by the server
// (jobs[].progress.steps or synthesized from importStatus.progress).
type JobStep struct {
	ID     string
	Name   string
	Status tui.StepState
}

// ImportJob mirrors the slice of Job the poller consumes (real k8s job or
// the pseudo-job Node synthesizes from importStatus.progress —
// status.ts:288-328; the synthesis lives in the command's fetch closure).
type ImportJob struct {
	CreatedAt   string
	CompletedAt string
	Status      string // progress.status; "" treated as "unknown" (status.ts:333)
	Steps       []JobStep
}

// FailedStep is a failed entry from importStatus.progress.steps
// (status.ts:361 failedImportStep).
type FailedStep struct {
	Name      string
	Output    []string
	StartedAt int64 // unix seconds
}

// ProgressSnapshot flattens one ImportSQLProgress response. Job == nil
// means "no job data available yet" — the poller waits (or fast-returns
// under ReturnMissingJobImmediately).
type ProgressSnapshot struct {
	Job                     *ImportJob
	StatusProgressStartedAt int64 // importStatus.progress.started_at (unix seconds)
	FailedStep              *FailedStep
	Launched                bool
}

// ProgressFetch retrieves the current snapshot (the command wraps
// gql.ImportSQLProgress).
type ProgressFetch func(ctx context.Context) (*ProgressSnapshot, error)

// CheckStatusOpts configures CheckStatus.
type CheckStatusOpts struct {
	Fetch    ProgressFetch
	Tracker  *tui.ProgressTracker
	Interval time.Duration
	// ReturnMissingJobImmediately — true for `vip import sql status`
	// (status.ts:198).
	ReturnMissingJobImmediately bool
	// OnPoll fires after each snapshot is applied to the tracker, before
	// terminal-state checks — the command renders its suffix block here.
	OnPoll func(createdAt, completedAt, overallStatus string)
}

// StatusResult is the terminal outcome of a finished (non-failed) poll.
type StatusResult struct {
	Status      string
	Message     string // e.g. "No import job found"
	CreatedAt   string
	CompletedAt string
}

// ImportFailedError ports ImportFailedError (status.ts:107).
type ImportFailedError struct {
	InImportProgress bool
	CommandOutput    []string
	ErrorText        string
	StepName         string
	Launched         bool
}

func (e *ImportFailedError) Error() string { return e.ErrorText }

// parseFlexibleTime mimics JS `new Date(s).getTime()`: accept the formats
// the API and the synthesis path produce. Returns ok=false for NaN cases.
func parseFlexibleTime(s string) (time.Time, bool) {
	for _, layout := range []string{
		time.RFC3339, time.RFC1123, time.RFC1123Z, time.RFC822, time.RFC850,
		"2006-01-02 15:04:05", "2006-01-02T15:04:05.000Z",
	} {
		if t, err := time.Parse(layout, s); err == nil {
			return t, true
		}
	}
	return time.Time{}, false
}

// CheckStatus ports importSqlCheckStatus's getResults loop
// (status.ts:267-417). The command owns rendering and exit codes; this
// owns the poll-state machine.
func CheckStatus(ctx context.Context, opts CheckStatusOpts) (*StatusResult, error) {
	interval := opts.Interval
	if interval == 0 {
		interval = DefaultPollInterval
	}
	overall := "Checking..." // status.ts:213

	for {
		snap, err := opts.Fetch(ctx)
		if err != nil {
			return nil, err
		}

		if snap.Job == nil {
			if opts.ReturnMissingJobImmediately {
				// status.ts:329 — resolve('No import job found')
				return &StatusResult{Message: "No import job found"}, nil
			}
			// status.ts:294 — progress meta not filled out yet; wait.
			select {
			case <-ctx.Done():
				return nil, ctx.Err()
			case <-time.After(interval):
			}
			continue
		}

		job := snap.Job
		jobStatus := job.Status
		if jobStatus == "" {
			jobStatus = "unknown" // status.ts:333
		}
		createdAt := job.CreatedAt
		completedAt := job.CompletedAt

		// failedImportStep gate (status.ts:353-366): the import_progress
		// meta is only pertinent when it started at/after job creation.
		var failed *FailedStep
		if jobCreation, ok := parseFlexibleTime(createdAt); ok &&
			snap.StatusProgressStartedAt*1000 >= jobCreation.UnixMilli() {
			if fs := snap.FailedStep; fs != nil && fs.StartedAt*1000 > jobCreation.UnixMilli() {
				failed = fs
			}
		}

		if len(job.Steps) == 0 {
			// status.ts:368 — reject({error: 'Could not enumerate the
			// import job steps'})
			return nil, errors.New("Could not enumerate the import job steps")
		}

		if failed != nil {
			// status.ts:373 — demote the 'import' step to failed, render,
			// then reject with the structured error.
			steps := make([]JobStep, len(job.Steps))
			copy(steps, job.Steps)
			for i := range steps {
				if steps[i].ID == "import" {
					steps[i].Status = tui.StepFailed
				}
			}
			opts.Tracker.SetStepsFromServer(toServerSteps(steps))
			if opts.OnPoll != nil {
				opts.OnPoll(createdAt, completedAt, "failed")
			}
			return nil, &ImportFailedError{
				InImportProgress: true,
				CommandOutput:    failed.Output,
				ErrorText:        "Import step failed",
				StepName:         failed.Name,
				Launched:         snap.Launched,
			}
		}

		opts.Tracker.SetStepsFromServer(toServerSteps(job.Steps))
		if opts.OnPoll != nil {
			opts.OnPoll(createdAt, completedAt, overall)
		}

		if jobStatus == "error" {
			// status.ts:399 — reject({error: 'Import job failed', ...})
			return nil, errors.New("Import job failed")
		}

		if jobStatus != "running" && completedAt != "" {
			// status.ts:404 — resolve(importJob)
			return &StatusResult{
				Status: jobStatus, CreatedAt: createdAt, CompletedAt: completedAt,
			}, nil
		}

		overall = "running" // status.ts:408

		select {
		case <-ctx.Done():
			return nil, ctx.Err()
		case <-time.After(interval):
		}
	}
}

func toServerSteps(steps []JobStep) []tui.ServerStep {
	out := make([]tui.ServerStep, len(steps))
	for i, s := range steps {
		out[i] = tui.ServerStep{Name: s.Name, Status: s.Status}
	}
	return out
}

// GetErrorMessage ports getErrorMessage (status.ts:116). Message blocks
// are copied verbatim, including blank lines and the conditional
// rollback notice (suppressed for launched environments).
func GetErrorMessage(fe *ImportFailedError) string {
	rollbackMessage := ""
	if !fe.Launched {
		rollbackMessage = "Your site is " + color.BlueString("automatically being rolled back") +
			" to the last backup prior to your import job.\n"
	}

	message := fe.ErrorText
	if !fe.InImportProgress {
		return message
	}

	commandOutputBlock := func() string {
		if fe.CommandOutput != nil {
			joined := strings.Join(fe.CommandOutput, ";")
			return "\nPlease inspect your input file and make the appropriate corrections before trying again.\nThe server said:\n> " +
				color.RedString(joined) + "\n"
		}
		return ""
	}

	switch fe.StepName {
	case "import_preflights":
		message += "\nThis error occurred prior to the mysql batch script processing of your SQL file.\n\nYour site content was not altered.\n\nIf this error persists, please contact support.\n"
	case "importing_db":
		message += "\nThis error occurred during the mysql batch script processing of your SQL file.\n\n" + rollbackMessage
		if fe.CommandOutput != nil {
			message += commandOutputBlock()
		} else {
			message += "Please contact support and include this message along with your sql file."
		}
	case "validating_db":
		message += "\nThis error occurred during the post-import validation of the imported data.\n\n" + rollbackMessage + "\n"
		if fe.CommandOutput != nil {
			message += commandOutputBlock()
		} else {
			message += "Please contact support and include this message along with your sql file."
		}
	case "update_primary_domain":
		message += "\nThis error occurred during the update of the primary domain.\n\n" + rollbackMessage + "\n"
		if fe.CommandOutput != nil {
			message += commandOutputBlock()
		}
	}
	return message
}

// Capitalize ports format.ts capitalize (format.ts:139).
func Capitalize(s string) string {
	if s == "" {
		return ""
	}
	return strings.ToUpper(s[:1]) + s[1:]
}
